sql/core/src/main/java/org/apache/spark/sql/sources/v2/BatchWriteSupportProvider.java

```diff
@@ -21,33 +21,39 @@
 
 import org.apache.spark.annotation.InterfaceStability;
 import org.apache.spark.sql.SaveMode;
-import org.apache.spark.sql.sources.v2.writer.DataSourceWriter;
+import org.apache.spark.sql.sources.v2.writer.BatchWriteSupport;
 import org.apache.spark.sql.types.StructType;
 
 /**
  * A mix-in interface for {@link DataSourceV2}. Data sources can implement this interface to
- * provide data writing ability and save the data to the data source.
+ * provide data writing ability for batch processing.
+ *
+ * This interface is used when end users want to use a data source implementation directly, e.g.
+ * {@code Dataset.write.format(...).option(...).save()}.
  */
 @InterfaceStability.Evolving
-public interface WriteSupport extends DataSourceV2 {
+public interface BatchWriteSupportProvider extends DataSourceV2 {
 
   /**
-   * Creates an optional {@link DataSourceWriter} to save the data to this data source. Data
+   * Creates an optional {@link BatchWriteSupport} to save the data to this data source. Data
    * sources can return None if there is no writing needed to be done according to the save mode.
    *
    * If this method fails (by throwing an exception), the action will fail and no Spark job will be
    * submitted.
    *
-   * @param writeUUID A unique string for the writing job. It's possible that there are many writing
-   *                  jobs running at the same time, and the returned {@link DataSourceWriter} can
-   *                  use this job id to distinguish itself from other jobs.
+   * @param queryId A unique string for the writing query. It's possible that there are many
+   *                writing queries running at the same time, and the returned
+   *                {@link BatchWriteSupport} can use this id to distinguish itself from others.
    * @param schema the schema of the data to be written.
    * @param mode the save mode which determines what to do when the data are already in this data
    *             source, please refer to {@link SaveMode} for more details.
    * @param options the options for the returned data source writer, which is an immutable
    *                case-insensitive string-to-string map.
-   * @return a writer to append data to this data source
+   * @return a write support to write data to this data source.
    */
-  Optional<DataSourceWriter> createWriter(
-      String writeUUID, StructType schema, SaveMode mode, DataSourceOptions options);
+  Optional<BatchWriteSupport> createBatchWriteSupport(
+      String queryId,
+      StructType schema,
+      SaveMode mode,
```

Comment:
I don't think this is a good idea. Why introduce a legacy API into a new API? If we are moving old sources to the new API, then they should fully implement the new API and should not continue to expose the unpredictable v1 behavior.

That said, as long as the `TableCatalog` makes it in, I don't care what anonymous tables do because I don't intend for any of our sources to use this path.

Reply:
The problem here is, if we don't take `saveMode`, the only end-user API to write to a data source is: `df.write.format(...).mode("append").save()`. That makes data source v2 totally unusable before we introduce the new write APIs.

I hope we can get this in before Spark 2.4, so that some data source projects can start migrating and experimenting.

Reply:
@cloud-fan: this is a bad idea **because** it enables save modes other than append for DataSourceV2 without using the new logical plans. This leads to undefined behavior and is why we proposed standard logical plans in the first place.

Using a data source implementation directly should only support appending and scanning anything more complex must require `DeleteSupport` (see #21308) or `TableCatalog` (see #21306) and the new logical plans. Otherwise, this will allow sources to expose behavior that we are trying to fix.

I'm -1 on this.

Reply:
To clarify, your proposal is that we should block the completion of DataSourceV2 until the new logical plans are in place?

Reply:
I think that the v2 sources should only use the plans proposed in the SPIP. I also think that the v2 data source API should always tell the data source exactly what to do: for overwrite, what should be deleted and what should be added.

That doesn't block fixing the v2 API here and doesn't prevent anyone from using it. But it would prevent people from relying on undefined behavior that results from passing an ambiguous `SaveMode` to a source.

The only thing that would not be available by doing this is overwrite support by using the SaveMode, which isn't something anyone should rely on because it doesn't have defined behavior.

I understand that this may seem like it would block migration from the v1 API to the v2 API. If so, that's the right thing to do until we clearly define how v2 sources behave.

Reply:
I totally agree that `SaveMode` is a bad API which leads to undefined behavior. That's why we start a project to design new DDL logical plans and write APIs. However, I believe we had an agreement before that we can't remove existing APIs, so the `DataFrameWriter` and `SaveMode` will still be there in Spark. If I'm a data source developer, even I've implemented the new write APIs (assuming it's finished), I would still support `SaveMode` to attract more users. `DataFrameWriter` is a very widely used API and the end users may need a long time to migrate to the new write APIs. BTW, file source (without catalog) does have a clearly defined behavior regarding `SaveMode`, we should make it possible to migrate file source to data source v2.